Merging photonvision into Thor

SamueljWilson opened the pull request.

SamueljWilson committed:
Updated to WPILib 2024.2.1

SamueljWilson committed:
Installed photonLib

SamueljWilson committed:
Installed photonLib

SamueljWilson committed:
Added the camera subsystem and all of the constants needed

SamueljWilson committed:
Added the camera subsystem and all of the constants needed

SamueljWilson committed:
Updated the odometry to include vision, and updated the robot container

SamueljWilson committed:
Updated the odometry to include vision, and updated the robot container

jasone committed:
Update to WPILib 2024.2.1

jasone committed:
Simplify/correct drive encoder conversion factors

The SPARK MAX controller uses native position units of motor revolutions and native velocity units of motor revolutions per minute ("RPM" in the documentation). As a result, we don't need to worry about encoder counts, nor do we need to keep track of conversion factors once they have been applied to the encoder. We can consistently work in meters (position) and meters per second (velocity) without repeating conversions.

SamueljWilson committed:
Updated the apriltag field layout to use the 2024 season layout

SamueljWilson committed:
Updated the apriltag field layout to use the 2024 season layout

SamueljWilson committed:
Updated photonvision

SamueljWilson committed:
Updated photonvision

SamueljWilson committed:
Updated the drive subsystem to take in the camera subsystem as one of its arguments

SamueljWilson committed:
Updated the drive subsystem to take in the camera subsystem as one of its arguments

SamueljWilson committed:
Changed the trust values for the swerve drive pose estimator

SamueljWilson committed:
Changed the trust values for the swerve drive pose estimator

SamueljWilson committed:
Cleaned up code

SamueljWilson committed:
Cleaned up code

Commit:
set up the limelight to turn towards note

Commit:
added SendableChooser for autonomous

Commit:
add removed auto to chooser

SamueljWilson committed:
Renamed cameras

SamueljWilson committed:
Renamed cameras

SamueljWilson committed:
Set the photon estimator tag model to 36h11

SamueljWilson committed:
Set the photon estimator tag model to 36h11

SamueljWilson committed:
Changed the loop that iterates over the different camera estimators results to only call the getFieldRelativePoseEstimators once

SamueljWilson committed:
Changed the loop that iterates over the different camera estimators results to only call the getFieldRelativePoseEstimators once

jasone commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -99,6 +99,7 @@ private SwerveModulePosition[] getPositions() {
 
   /** Creates a new DriveSubsystem. */
   public DriveSubsystem(CameraSubsystem cameraSubsystem) {
+    m_photonCameras = cameraSubsystem;
```
The name `m_photonCameras` suggests this is something more specific than the camera subsystem (confusing).

jasone commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -95,7 +95,7 @@ private SwerveModulePosition[] getPositions() {
         m_initialPose,
         VecBuilder.fill(0.005, 0.005, Math.toRadians(1)),
-        VecBuilder.fill(0.05, 0.05, Math.toRadians(5)));
+        VecBuilder.fill(0.005, 0.005, Math.toRadians(1)));
```
We're going to want named constants for these.

jasone commented on src/main/java/frc/robot/subsystems/DriveSubsystem.java:
```diff
@@ -131,16 +132,17 @@ public void periodic() {
-      Optional<EstimatedRobotPose> robotPoseEstimator = m_photonCameras.getFieldRelativePoseEstimators().get(i);
+    
+    ArrayList<Optional<EstimatedRobotPose>> photonRobotPoseList = m_photonCameras.getFieldRelativePoseEstimators();
+    for (int i=0; i<photonRobotPoseList.size(); i++) {
```
This could be written as:

```java
photonRobotPoseList.forEach(robotPoseEstimator -> {
  [...]
});
```

Or, as long as the stuff in front of `forEach` isn't too complicated, and you don't need `photonRobotPoseList` for multiple operations:

```java
m_photonCameras.getFieldRelativePoseEstimators().foreach(robotPoseEstimator -> {
  [...]
});
```

It's usually a good idea to use "for each"-style iteration rather than indexed iteration whenever you don't explicitly need the index inside the loop, because there are fewer ways to mess up the boilerplate code.

jasone commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -148,4 +150,54 @@ public static final class AutoConstants {
+    public static final Transform3d kCamera2ToRobotOffset = new Transform3d(0.0, -0.12, 0.0, new Rotation3d());
+
+
+    public static final double kCamera1HeightMeters = 0.215;
+    public static final double kCamera2HeightMeters = 0.0;
+    public static final double kCamera3HeightMeters = 0.0;
+
+    public static final double KCameraPitchRadians = 0.0;
+
+    public static final double kTarget1HeightMeters = 0.540;
+    public static final double kTarget2HeightMeters = 0.522;
+    public static final double kTarget3HeightMeters = 0.530;
+    public static final double kTarget4HeightMeters = 0.0;
+    public static final double kTarget5HeightMeters = 0.0;
+    public static final double kTarget6HeightMeters = 0.0;
+    public static final double kTarget7HeightMeters = 0.0;
+    public static final double kTarget8HeightMeters = 0.0;
```
A later diff removes uses of these constants, but the constants appear to remain, no longer used.

jasone reviewed (changes requested):
Let's get the constants cleaned up before merging this pull request. Up to you on whether you address the other feedback before merging.

SamueljWilson committed:
Renamed m_photonCameras to m_cameraSystem

SamueljWilson committed:
Renamed m_photonCameras to m_cameraSystem

SamueljWilson committed:
Added the standard deviations as constants

SamueljWilson committed:
Added the standard deviations as constants

SamueljWilson committed:
Removed constants not needed

SamueljWilson committed:
Removed constants not needed

SamueljWilson committed:
Removed command that would never be used

SamueljWilson committed:
Removed command that would never be used

SamueljWilson committed:
Changed the periodic to use forEach instead of a for loop

SamueljWilson committed:
Changed the periodic to use forEach instead of a for loop

SamueljWilson committed:
Renamed the camera subsystem to m_cameraSystem and removed  unnecessary dependency

SamueljWilson committed:
Renamed the camera subsystem to m_cameraSystem and removed  unnecessary dependency

SamueljWilson committed:
Updated settings

Commit:
added constants

jasone commented on .vscode/settings.json:
```diff
@@ -25,5 +25,6 @@
   ],
-  "java.test.defaultConfig": "WPIlibUnitTests"
+  "java.test.defaultConfig": "WPIlibUnitTests",
+  "java.jdt.ls.vmargs": "-XX:+UseParallelGC -XX:GCTimeRatio=4 -XX:AdaptiveSizePolicyWeight=90 -Dsun.zip.disableMemoryMapping=true -Xmx2G -Xms100m -Xlog:disable"
```
Where did this come from?

jasone committed:
Treat acceleration/deceleration separately

Refactor `TrapezoidalConstraint` to treat acceleration and deceleration separately, and set the initial deceleration to 15 m/s^2. The robot is physically capable of slowing down more rapidly than speeding up, and tuning these parameters separately may give a more natural feel when driving, e.g. making it easier to rapidly approach a stopping position.

jasone committed:
Separate max robot/wheel velocity

It's okay for a wheel to move faster than the linear robot velocity, and allowing this makes the robot move more naturally when spinning.

SamueljWilson committed:
Merge branch 'Merging-photonvision' of https://github.com/SamueljWilson/Thor_2023 into Merging-photonvision

SamueljWilson committed:
Deleted debugging code

SamueljWilson committed:
Removed vscode setting

jasone commented:
Obsoleted by #39.
